Source/santametricservice/Writers/SNTMetricHTTPWriter.m

```diff
@@ -0,0 +1,97 @@
+/// Copyright 2021 Google Inc. All rights reserved.
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///    http://www.apache.org/licenses/LICENSE-2.0
+///
+///    Unless required by applicable law or agreed to in writing, software
+///    distributed under the License is distributed on an "AS IS" BASIS,
+///    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+///    See the License for the specific language governing permissions and
+///    limitations under the License.
+#include <dispatch/dispatch.h>
+#import <MOLAuthenticatingURLSession/MOLAuthenticatingURLSession.h>
+
+#import "Source/santametricservice/Writers/SNTMetricHTTPWriter.h"
+
+@implementation SNTMetricHTTPWriter {
+ @private
+  NSMutableURLRequest *_request;
+  MOLAuthenticatingURLSession *_authSession;
+  NSURLSession *_session;
+}
+
+- (instancetype)init {
+  self = [super init];
+  if (self) {
+    _request = [[NSMutableURLRequest alloc] init];
+    _request.HTTPMethod = @"POST";
+    _authSession = [[MOLAuthenticatingURLSession alloc] init];
+  }
+  return self;
+}
+
+/**
+ * Post serialzied metrics to the specified URL one object at a time.
+ **/
+- (BOOL)write:(NSArray<NSData *> *)metrics toURL:(NSURL *)url error:(NSError **)error {
+  // open the file and write it.
+  __block NSError *_blockError = nil;
+
+  NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:url];
+  request.HTTPMethod = @"POST";
+  [request setValue:@"application/json" forHTTPHeaderField:@"Content-Type"];
+
+  _authSession.serverHostname = url.host;
+  _session = _authSession.session;
+
+  dispatch_semaphore_t semaphore = dispatch_semaphore_create(0);
+  dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
```

Comment:
What's the reasoning for this dispatch_async? Each call to `[dataTaskWithRequest:completionHandler: resume]` should be async anyway so the semaphore here is signalling only after the call to start the request occurs.